### PR TITLE
feat: Add functions for downloading artifacts

### PR DESCRIPTION
**Pull request by tonyo**

### CODE REVIEW

File: src/__tests__/transport.test.ts [hunk 1]
```diff
@@ -124,6 +127,30 @@ describe('Transport', () => {
...
+        return Promise.resolve(resp);
+      });
+
+      const response = await transport.request('something');
+      expect(response).toEqual({ some: 'data' });
+    });
+
+    test('resolves undefined for status 204', async () => {
+      const transport = new Transport();
+      requestMock.mockImplementation(() => {
```
Comment (jan-auer) [inline src/__tests__/transport.test.ts]: Make this `async` and simply return the response.

File: src/client.ts [hunk 2]
```diff
@@ -54,14 +56,23 @@ export interface JobOptions {
 export class Client {
   /** Internal transport used to send requests to Zeus. */
   private readonly transport: Transport;
+  /** GitHub repository owner */
+  public readonly owner: string;
+  /** GitHub repository name */
+  public readonly repo: string;
```
Comment (jan-auer) [inline src/client.ts]: I'd prefer to not have `owner` / `repo` as part of the client configuration, but rather options in the methods.

File: src/client.ts [hunk 3]
```diff
@@ -54,14 +56,23 @@ export interface JobOptions {
...
+  /** Path to directory where downloaded files are stored */
+  public readonly downloadDirectory: string;
```
Comment (jan-auer) [inline src/client.ts]: This should be called something like `tempDirectory` or `defaultDirectory`. Usually, the application will want to use explicit download locations every time it uses `downloadArtifact`, so it's better to clearly distinguish.

File: src/client.ts [hunk 4]
```diff
@@ -183,4 +194,146 @@ export class Client {
...
+   * The file is placed in the download directory. If the file does not exist,
+   * an error is thrown.
+   *
+   * @param artifact A file object to download
+   * @returns Absolute path to the local copy of the file
+   * @async
+   */
+  public async downloadArtifact(
+    artifact: Artifact,
+    downloadDirectory?: string
```
Comment (jan-auer) [inline src/client.ts]: _nit_: `downloadDirectory?: string = this.downloaddirectory`

File: src/client.ts [hunk 5]
```diff
@@ -183,4 +194,146 @@ export class Client {
...
+  ): Promise<string> {
+    const downloadDir = downloadDirectory || this.downloadDirectory;
+    if (!downloadDir) {
+      throw new Error('Download directory not specified');
+    }
+
+    const url = this.getUrl(artifact.download_url);
```
Comment (jan-auer) [inline src/client.ts]: _nit_: We could add a check if the directory exists and throw if not. However, applications should usually use something like `mkdirp` to create the path on the fly.

File: src/client.ts [hunk 6]
```diff
@@ -183,4 +194,146 @@ export class Client {
...
+
+  /**
+   * Downloads a file from the store
+   *
+   * The file is placed in the download directory. If the file does not exist,
+   * an error is thrown.
+   *
+   * @param artifact A file object to download
+   * @returns Absolute path to the local copy of the file
+   * @async
```
Comment (jan-auer) [inline src/client.ts]: _nit_: The `@async` tags are redundant, as this is declared in the typescript signature.

File: src/transport.ts [hunk 7]
```diff
@@ -29,6 +30,12 @@ export interface Options {
   token?: string;
   /** A logger with the same interface like console. */
   logger?: Logger;
+  /** Repository owner */
+  owner?: string;
+  /** Repository name */
+  repo?: string;
+  /** Download directory */
+  downloadDirectory?: string;
```
Comment (jan-auer) [inline src/transport.ts]: These options certainly do not belong in here. You could re-export another `Options` interface from the client like this:

```ts
import { Options as TransportOptions } from './transport';

export interface Options extends TransportOptions {
  /** Repository owner. */
  owner?: string;
  /** Repository name. */
  repo?: string;
  /** Default download directory. */
  downloadDirectory?: string;
}
```

File: src/transport.ts [hunk 8]
```diff
@@ -95,37 +102,59 @@ export class Transport {
    * a fully qualified URL with protocol and host. In that case, the request is
    * performed to the path as is.
    *
+   *
```
Comment (jan-auer) [inline src/transport.ts]: _nit_: Remove second newline.

**Review by jan-auer — CHANGES_REQUESTED** [review]

File: src/client.ts (hunk 2 above)
Comment (tonyo) [reply src/client.ts]: You're right: this is an sdk lib, so better to make things explicit.

File: src/client.ts (hunk 5 above)
Comment (tonyo) [reply src/client.ts]: Added some additional checks that the path exists and that it's a directory.

File: src/transport.ts (hunk 7 above)
Comment (tonyo) [reply src/transport.ts]: 👍 cool stuff

**Comment by tonyo** [comment]
@jan-auer ready for another pass!

**Pull request merged**